### PR TITLE
Support untagged union

### PR DESCRIPTION
**Pull request by daddykotex**
Adds support for untagged union as well as a bit of documentation regarding union encoding/decoding.

### CODE REVIEW

File: modules/protocol/resources/META-INF/smithy/smithy4s.smithy [hunk 1]
```diff
@@ -56,5 +56,8 @@ structure UncheckedExample {
 structure uuidFormat {
 }
 
-@trait(selector: "union")
+@trait(selector: "union :not([trait|smithy4s.api#untagged])")
```
Comment (Baccata) [inline modules/protocol/resources/META-INF/smithy/smithy4s.smithy:59]: 👍

**Review by lewisjkl — APPROVED** [review]

File: modules/core/src/smithy4s/internals/Hinted.scala [hunk 2]
```diff
@@ -138,6 +138,21 @@ object Hinted {
         f: (Option[H1], Option[H2]) => F[A]
     ): Hinted[F, A] =
       Hinted(Hints(), (h: Hints) => f(h.get[H1], h.get[H2]))
+    def onHintOneOf[H1: ShapeTag, H2: ShapeTag, A](
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/Hinted.scala:141]: Interesting. I'll allow it 😄

File: modules/docs/src/03-protocols/02-simple-rest-json/01-overview.md [hunk 3]
```diff
@@ -53,3 +67,83 @@ This protocol and its interpreters, are aware of the following traits provided o
 * [error trait](https://awslabs.github.io/smithy/1.0/spec/core/type-refinement-traits.html#error-trait)
 * [http traits](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html), including **http**, **httpError**, **httpLabel**, **httpHeader**, **httpPayload**, **httpQuery**, **httpPrefixHeaders**, **httpQueryParams**.
 * [timestampFormat trait](https://awslabs.github.io/smithy/1.0/spec/core/protocol-traits.html?highlight=timestampformat#timestampformat-trait)
+
+## Decoding and encoding unions
```
Comment (Baccata) [inline modules/docs/src/03-protocols/02-simple-rest-json/01-overview.md:71]: 👍  awesome

**Review by Baccata — APPROVED** [review]
Some nitpick regarding name and placement of a method, consider it mergeable after addressed

File: modules/core/src/smithy4s/internals/Hinted.scala (hunk 2 above)
Comment (kubukoz) [reply modules/core/src/smithy4s/internals/Hinted.scala:141]: personally I like to do sth like 

```scala

Hinted[F].from { hints =>
  H1(hints).map(onH1)
    ..orElse(H2(hints).map(onH2))
}
```

but there's still the `from` outside that may not be the most convenient :)

Comment (Baccata) [reply modules/core/src/smithy4s/internals/Hinted.scala:141]: Maybe we should generate custom extractors, and allow for : 

``` 
Hinted[F].from { 
    case H1(...) => onH1
    case H2(...) => onH2
} 
```

Comment (daddykotex) [reply modules/core/src/smithy4s/internals/Hinted.scala:141]: I tried a top level extractor in the companion object directly but it conflicts with case class default extractors. instead, I went for something that produce this:

```scala
package smithy4s.api

import smithy4s.syntax._

case class Untagged()
object Untagged extends smithy4s.ShapeTag.Companion[Untagged] {
  val id: smithy4s.ShapeId = smithy4s.ShapeId("smithy4s.api", "untagged")

  val hints : smithy4s.Hints = smithy4s.Hints(
    id,
    smithy.api.Trait(Some("union :not([trait|smithy4s.api#discriminated])"), None, None),
  )

  val schema: smithy4s.Schema[Untagged] = constant(Untagged()).withHints(hints)
  implicit val staticSchema : schematic.Static[smithy4s.Schema[Untagged]] = schematic.Static(schema)
  
  object extractHint { def unapply(h: smithy4s.Hints): Option[Untagged] = h.get[Untagged] }
}
```

wdyt?

Comment (daddykotex) [reply modules/core/src/smithy4s/internals/Hinted.scala:141]: see https://github.com/disneystreaming/smithy4s/pull/178/commits/778ac7296b55af18cb292688d73985cccf8f881f

Comment (daddykotex) [reply modules/core/src/smithy4s/internals/Hinted.scala]: got rid of it in favor of extractors for `ShapeTag.Companion` and `NewType`

**Review by Baccata — APPROVED** [review]
Great work

**Pull request merged**